KEP-2862: Fine-grained Kubelet API Authorization (#4760)
kubernetes/enhancements: keps/sig-auth/2862-fine-grained-kubelet-authz/README.md

[kubernetes/website]: https://git.k8s.io/website

## Summary

<!--
This section is incredibly important for producing high-quality, user-focused
documentation such as release notes or a development roadmap. It should be
possible to collect this information before implementation begins, in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself. KEP editors and SIG Docs
should help to ensure that the tone and content of the `Summary` section is
useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

We propose a change to the Kubelet API authorization to provide more
fine-grained control. Currently, the API uses a coarse authorization scheme,
where actions like reading health status and the ability to exec into a pod

> require the same permission.

Reviewer: AIUI, this is a primarily change to the Node API. Does that sound right?

Author: It is a change to the kubelet API. So agents (think logging and monitoring) that interact with the kubelet directly.

Reviewer: Got it. Please make that clear so that people just reading the summary can tell what we propose to change.

You could also mention the intent to document the existing stable kubelet API.

Author: Thanks! Updated.

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

The Kubelet API on the authenticated(10250) port currently uses a very coarse
scheme for authorizing requests. For example, reading `/healthz` and calling
`/exec/…` (i.e. execute arbitrary code) require the same `proxy` subresource.
So, if an application needs to say `list` `pods` on the node, as many monitoring
and logging agents do, they must be granted the `proxy` subresource in RBAC.
Doing so grants many other powerful permissions to these agents which could be
exploited by an attacker to escalate privilege.

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

> - Introduce new subresources for `/healthz` and `/pods/` kubelet endpoints,
> allowing for more  granular authorization.

Reviewer: Is this right?
```suggestion
- Introduce new subresources for a Node, `healthz` and `pods`, that are proxied to the kubelet endpoints
 `/healthz` and `/pods/` respectively. Making this change allowsfor more  granular authorization.
```

Reviewer 2: As commented below, these are subresources for the sake of the `SubjectAccessReview` issued by the Kubelet, but these aren't really Kubernetes REST API subresources. These are more like non-resource URLs, and are only exposed in the Kubernetes API through the proxy subresource to the node. The authorization being considered here is only relevant when accessing the Kubelet endpoint directly.

Author: Updated to say authorization subresource instead to disambiguate like you suggested. Thanks!

you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

Add new subresources for `healthz` and `pods` endpoint while supporting the old coarse grained
proxy subresources as shown in the table below.

| Request path                                     | Existing Subresource | Proposed Subresource(s) |
| ----------------------------------------------- | -------------------- | ------------------------ |
| /healthz                                        | proxy                | proxy, healthz           |
| /healthz/log                                    | proxy                | proxy, healthz           |
| /healthz/ping                                   | proxy                | proxy, healthz           |
| /healthz/syncloop                               | proxy                | proxy, healthz           |
| /pods/                                          | proxy                | proxy, pods              |
| /runningpods/                                   | proxy                | proxy, pods              |

> This way users who were previously using the `proxy` verb to access `/healthz`
> kubelet endpoint can update their `ClusterRole` or `Role` by replacing the
> `proxy` verb with the `healthz` verb. 

Reviewer: ```suggestion
This way users who were previously using the `proxy` subresource to grant access to the
`/healthz` kubelet endpoint can update their ClusterRole or Role by replacing the
`proxy` subresource with the `healthz` subresource.
```
?

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

As a security conscious node monitoring agent owner I want to interact with the
kubelet API to list pods on the node without having to grant the agent the `proxy` subresource and use
least privilege to grant the exact permissions the agent needs like `pods` or 
`healthz`.

### Notes

If this change were to be implemented as proposed, the `proxy` `subresource` 
would still cover `/attach/`, `/exec/`, `/run/`, `/configz/`, `/debug/`. 
The reader might be wondering why we didn't break these permissions out into 

> their own `subresource`?

Reviewer: (nit)
```suggestion
their own subresource?
```

The reason why we did not breakout all endpoints under `proxy` into their own

> `subresource` is because we could not reason about a case where if we allowed

Reviewer: (nit)
```suggestion
subresource is because we could not reason about a case where if we allowed
```

one of these permissions, having any of the other permissions would be considered
worse. If you have `/exec/` then having `/run/` or `/attach/` isn't making it 
worse for an attacker.

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

> Since we are adding a second `SubjectAccessReview` request the latency for some
> requests will increase. However, `kubelet` has a `SubjectAccessReview` request
> cache, so subsequent requests should result in cache hits.

Reviewer: ```suggestion
Since we are adding a second SubjectAccessReview request the latency for some
requests will increase. However, `kubelet` has a SubjectAccessReview` response
cache, so subsequent requests should result in cache hits.
```

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

> To determine if the caller has the required permissions for a particular request
> kubelet creates a `SubjectAccessReview` request to `kube-apiserver`. The
> `SubjectAccessReviews` are currently populated with:

Reviewer: ```suggestion
To determine if the caller has the required permissions for a particular request
kubelet creates a SubjectAccessReview request to `kube-apiserver`. The
SubjectAccessReviews are currently populated with:
```

```yaml
apiVersion: authorization.k8s.io/v1
kind: SubjectAccessReview
spec:
  user: user1
  uid: 64167384-10aa-4361-bcef-526ab51d1e4d
  groups:
  - groups1
  resourceAttributes:
    group: ""
    version: "v1"
    resource: "nodes"
    namespace: ""
    name: "node-1"
    subresource: "proxy"
    verb: "GET"
```

> The following information is passed into the `SubjectAccessReview` request
> - Requesting user (generally from a `TokenReviewRequest` or client certificate authentication)

Reviewer: ```suggestion
The following information is passed into the SubjectAccessReview request
- Requesting user (generally from a TokenReviewRequest or client certificate authentication)
```

The subresource is determined by the path of the request. `kube-apiserver` upon
receiving this request will check if the user in the `SubjectAccessReview` has
the RBAC permissions. The RBAC for the sameple request above would look like as
follows:-

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: nodes-proxy
rules:
- apiGroups: [""]
  resources: ["nodes/proxy"]
  verbs: ["get", "create"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: nodes-proxy-global
subjects:
- apiGroup: rbac.authorization.k8s.io
  kind: User
  name: user1
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: nodes-proxy
```

We will add a new feature-gate `KubeletFineGrainedAuthz` that will be defaulted to
`false` in `alpha`. When this feature-gate is set to `true` kubelet will first send
a `SubjectAccessReview` using the fine-grained verb for `healthz` and `pods` endpoints.
If that request fails kubelet will retry with the coarse-grained verb (`proxy`).

When `kube-apiserver` communicates with the kubelet
its user is bound to the `system:kubelet-api-admin` role. When the feature-gate 

> is enabled we will update the `ClusterRole` as follow:-

Author: Updated. thanks!

```diff
  apiVersion: rbac.authorization.k8s.io/v1
  kind: ClusterRole
  metadata:
    annotations:
      rbac.authorization.kubernetes.io/autoupdate: "true"
    labels:
      kubernetes.io/bootstrapping: rbac-defaults
    name: system:kubelet-api-admin
  rules:
  - apiGroups:
    - ""
    resources:
    - nodes
    verbs:
    - get
    - list
    - watch
  - apiGroups:
    - ""
    resources:
    - nodes
    verbs:

>     - proxy

Reviewer: This seems to be allowing full access to proxy to all nodes. Is that what we want to demonstrate?

Author: This role already exists and has full access to proxy to all nodes. I just want to demonstrate the additional permissions we will add to this role.

extending the production code to implement this enhancement.
-->

- The following unit tests will be added to `nodeAuthorizerAttributesGetter`
  - When feature gate is disabled for `/healthz` and `/pods/` only 1 
  attribute with `proxy` verb should be returned
  - When feature gate is enabled for `/healthz` and `/pods/` 2 attributes should
  be returned and the first should use verb `healthz` or `pods` and the second should use 
  the `proxy` verb.

##### Integration tests

<!--
Integration tests are contained in k8s.io/kubernetes/test/integration.
Integration tests allow control of the configuration parameters used to start the binaries under test.
This is different from e2e tests which do not allow configuration of parameters.
Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
-->

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

> Well add the following integration tests:

Reviewer: (nit)
```suggestion
We'll add the following integration tests:
```

-->

###### Does this feature depend on any specific services running in the cluster?

<!--
Think about both cluster-level services (e.g. metrics-server) as well
as node-level agents (e.g. specific version of CRI). Focus on external or
optional services that are needed. For example, if this feature depends on
a cloud provider API, or upon an external software-defined storage or network
control plane.

For each of these, fill in the following—thinking about running existing user workloads
and creating new ones, as well as about cluster-level services (e.g. DNS):
  - [Dependency name]
    - Usage description:
      - Impact of its outage on the feature:
      - Impact of its degraded performance or high-error rates on the feature:
-->

### Scalability

<!--
For alpha, this section is encouraged: reviewers should consider these questions
and attempt to answer them.

For beta, this section is required: reviewers must answer these questions.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

###### Will enabling / using this feature result in any new API calls?

> <!--
> Describe them, providing:
>   - API call type (e.g. PATCH pods)
>   - estimated throughput
>   - originating component(s) (e.g. Kubelet, Feature-X-controller)
> Focusing mostly on:
>   - components listing and/or watching resources they didn't before
>   - API calls that may be triggered by changes of some Kubernetes resources
>     (e.g. update of object X triggers new updates of object Y)
>   - periodic API calls to reconcile state (e.g. periodic fetching state,
>     heartbeats, leader election, etc.)
> -->

Reviewer: (optionally)
```suggestion
For some requests, the API server performs an additional SubjectAccessReview (only when a check
for the `proxy` subresource wasn't authorized).
```

Reviewer 2: Yes, please note the additional SARs here.

Slight correction to @sftim's suggestion: SARs are sent by the Kubelet, and the proxy subresource is the second check.

Author: Update. Thanks!

-->

We propose a change to the Kubelet API authorization to provide more
fine-grained control. Currently, the API uses a coarse authorization scheme,
where actions like reading health status and the ability to exec into a pod
require the same permission.

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

The Kubelet API on the authenticated(10250) port currently uses a very coarse
scheme for authorizing requests. For example, reading `/healthz` and calling
`/exec/…` (i.e. execute arbitrary code) require the same `proxy` subresource.
So, if an application needs to say `list` `pods` on the node, as many monitoring
and logging agents do, they must be granted the `proxy` subresource in RBAC.
Doing so grants many other powerful permissions to these agents which could be

> exploited by an attacker to escalate privilege.

Reviewer 2: Might be worth adding the historical context: healthz and pods used to be availabel on the read-only port. With the removal of the read only port, it's raised the priority of this issue.

Author: Done. Thanks!

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->
- Introduce new subresources for `/healthz` and `/pods/` kubelet endpoints,
allowing for more  granular authorization.
- These changes should be backwards compatible.
- These changes should not break users on upgrade.

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->
- Create a new Kubelet API.
- Remove the existing subresources.
- Make the kubelet API node-restricted.

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

Add new subresources for `healthz` and `pods` endpoint while supporting the old coarse grained

> proxy subresources as shown in the table below.

Reviewer 2: "subresource" is a little overloaded here. These are represented as subresources in the `SubjectAccessReview`, but aren't really subresources in the sense of the Kubernetes REST API. To disambiguate, consider referring to these as "authorization subresource".

Author: updated subresource with "authorization subresource". Good call!

| Request path                                     | Existing Subresource | Proposed Subresource(s) |
| ----------------------------------------------- | -------------------- | ------------------------ |
| /healthz                                        | proxy                | proxy, healthz           |
| /healthz/log                                    | proxy                | proxy, healthz           |
| /healthz/ping                                   | proxy                | proxy, healthz           |
| /healthz/syncloop                               | proxy                | proxy, healthz           |
| /pods/                                          | proxy                | proxy, pods              |

> | /runningpods/                                   | proxy                | proxy, pods              |

Reviewer 3: I don't think we documented those endpoints or have any stability guarantees. Introducing special permissions for those make those endpoints official.

We need to discuss in SIG Node forum whether we want to document those endpoints.

@dchen1107 @mrunalp

Author: Where are the other endpoints documented? Can you point me to those?

Reviewer 3: just to update this thread: we will have a discussion on SIG Node meeting next week

Reviewer 4: cc/ @tallclair

Reviewer 2: I disagree that modifying these endpoints makes them any less official. I don't think we can use "unsupported by obscurity" here. If we want, we could include a warning on the documentation that these endpoints are not fully supported, and not subject to the same API guarantees. But in practice, I don't think we can remove or change these in the foreseeable future.

Reviewer 3: As discussed at SIG Node meeting, one of the goals of this KEP will be to document endpoints that were not documented before. It will eliminate this concern

Author: good call updated the KEP goals to include the documentation updates.

This way users who were previously using the `proxy` subresource to grant access to the
`/healthz` kubelet endpoint can update their ClusterRole or Role by replacing the
`proxy` subresource with the `healthz` subresource.

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

As a security conscious node monitoring agent owner I want to interact with the
kubelet API to list pods on the node without having to grant the agent the `proxy` subresource and use
least privilege to grant the exact permissions the agent needs like `pods` or 
`healthz`.

### Notes

If this change were to be implemented as proposed, the `proxy` `subresource` 

> would still cover `/attach/`, `/exec/`, `/run/`, `/configz/`, `/debug/`. 

Reviewer 3: what is the motivation to keep `/configz/`under `proxy`? Reading config maybe useful for monitoring tools

Reviewer 2: +1 for separating configz

Author: Done. Thanks!

The reader might be wondering why we didn't break these permissions out into 
their own subresource?

The reason why we did not breakout all endpoints under `proxy` into their own
subresource is because we could not reason about a case where if we allowed
one of these permissions, having any of the other permissions would be considered
worse. If you have `/exec/` then having `/run/` or `/attach/` isn't making it 

> worse for an attacker.

Reviewer 2: This seems like an argument for separating `configz`? That's also readonly right?

Author: Updated. Thanks!

We picked `/healthz` and `/pods` endpoints as they are read-only.

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

Since we are adding a second SubjectAccessReview request the latency for some
requests will increase. However, `kubelet` has a SubjectAccessReview response

> cache, so subsequent requests should result in cache hits.

Reviewer 2: This also increases the SAR QPS, but likewise this should be mitigated by the cache.

  - Subresource: which, before this KEP, can be one of: `proxy`, `log`, `metrics`, `spec`, `stats`

The subresource is determined by the path of the request. `kube-apiserver` upon
receiving this request will check if the user in the SubjectAccessReview is authorized
to access the relevant resource or subresource. For example, if a cluster is using RBAC
then this ClusterRole and ClusterRoleBinding might allow access:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: nodes-proxy
rules:
- apiGroups: [""]
  resources: ["nodes/proxy"]
  verbs: ["get", "create"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: nodes-proxy-global
subjects:
- apiGroup: rbac.authorization.k8s.io
  kind: User
  name: user1
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: nodes-proxy
```

We will add a new feature-gate `KubeletFineGrainedAuthz` that will be defaulted to
`false` in `alpha`. When this feature-gate is set to `true` kubelet will first send
a SubjectAccessReview specifically for the `healthz` or `pods` endpoint, based on the path of the
request made to the kubelet.

> If that request fails kubelet will retry with the coarse-grained verb (`proxy`).

Reviewer 2: Add a note about the cache here. I think the Kubelet should check the cache fore both resources before sending either request.

Another optimization to consider: if there is an expired cache entry for the proxy subresource, check that subresource first.

Author: Added a note about caching. I don't think the webhook authorizer which implements the cache allows that level of control over the requests. We may end up sending a request that would fail if the TTL of the deny cache is over. However the cache TTLs are pretty forgiving and this will only happen for requests that are infrequent. I want to avoid adding another layer of caching to keep the changes small, but we could add caching in the new Authorizer layer which calls into the webhook authorizer. WDYT?

Reviewer 2: Do we cache denials? If so, I think it's OK to leave it out, we can always optimize later. I might have been misremembering, thinking we don't cache denials...

Author: Yeah the webhook authorizer which is what kubelet uses and we won't change that does have cache for denials. https://github.com/kubernetes/kubernetes/blob/0aea469b96a1c0fed48dd2d4a0af96d36e24968c/staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go#L280C6-L280C17

you need any help or guidance.
-->

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.
-->

###### How can this feature be enabled / disabled in a live cluster?

<!--
Pick one of these and delete the rest.

Documentation is available on [feature gate lifecycle] and expectations, as
well as the [existing list] of feature gates.

[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
-->

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: `KubeletFineGrainedAuthz`
  - Components depending on the feature gate: 
    - `kubelet`
    - `kube-apiserver`
- [ ] Other
  - Describe the mechanism:
  - Will enabling / disabling the feature require downtime of the control
    plane?
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node?

###### Does enabling the feature change any default behavior?

<!--
Any change of default behavior may be surprising to users or break existing
automations, so be extremely careful here.
-->

> No, since we still send a SubjectAccessReview with the `proxy` subresource.

Reviewer 2: No change to user-facing behavior, but it does result in an extra SAR being sent.

Author: Updated. Thanks!

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

<!--
Describe the consequences on existing workloads (e.g., if this is a runtime
feature, can it break the existing applications?).

Feature gates are typically disabled by setting the flag to `false` and
restarting the component. No other changes should be necessary to disable the
feature.

NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
-->

> Yes.

Reviewer 2: A workload that is only authorized to use the new subresources will lose access.

Author: Updated, Thanks!

mid-rollout?

Be sure to consider highly-available clusters, where, for example,
feature flags will be enabled on some API servers and not others during the
rollout. Similarly, consider large clusters and how enablement/disablement
will rollout across nodes.
-->

###### What specific metrics should inform a rollback?

<!--
What signals should users be paying attention to when the feature is young
that might indicate a serious problem?
-->

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

<!--
Describe manual testing that was done and the outcomes.
Longer term, we may want to require automated upgrade/rollback tests, but we
are missing a bunch of machinery and tooling and can't do that now.
-->

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

<!--
Even if applying deprecation policies, they may still surprise some users.
-->

### Monitoring Requirements

<!--
This section must be completed when targeting beta to a release.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.

> -->

Reviewer 2: We already have metrics tracking request counts to Kubelet endpoints. I don't think we need to add a metric to track which subresource was used to authorize the request.

Is there a generic metric tracking client requests from the Kubelet that would capture the rate of SARs being sent?

Author: I didn't find any metrics for kubelet where we could track its client requests but could we rely on apiserver_request_total? While it won't let us distinguish between the callers one would still be able to capture the rate of incoming SARs to kube-apiserver.

Reviewer 2: SGTM.

-->

Historically the `/healthz` and 
`/pods` endpoints were available on the unauthenticated kubelet read-only port(10255) 
but kubelet read-only port is disabled by default and enabling the port is
considered a security worst practice. This has caused a lot of monitoring and 
logging agents to switch to using the kubelet authenticated port (10250).
However, the Kubelet API on the authenticated(10250) port currently uses a very
coarse scheme for authorizing requests. For example, reading `/healthz` and 
calling `/exec/…` (i.e. execute arbitrary code) require the same `proxy` 
subresource. So, if an application needs to say `list` `pods` on the node, as 
many monitoring and logging agents do, they must be granted the `proxy` 
subresource in RBAC. Doing so grants many other powerful permissions to these 
agents which could be exploited by an attacker to escalate privilege.  

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->
- Introduce new authorization subresources for `configz`, `/healthz` and `/pods/` kubelet endpoints,
allowing for more  granular authorization.

> - Officially document the `configz`, `healthz` and `/pods/` endpoints.

Reviewer: Aside: we should document the kubelet anyway; you don't need a KEP to document existing stable features.

Author: In sig node we discussed making this a part of the goals of this KEP. These endpoints were previously undocumented and they suggested that we document them if we are going to add fine-grained authorization for these endpoints. So I'll leave them as a goal of this KEP.
